### DISCUSSION TITLE
[ALG 101] LIOJ 1018 大平台

### DISCUSSION THREAD

**Discussion by PT-HUANG** [Lidemy/forum#302] [其他問題]
老師好，各位同學好
想請問LIOJ 1018 大平台的解法問題
解題邏輯如下：
1. 建立一個result空陣列
2. 設定index = 0
3. 讀取數字
4. 如果相同則跳過
5. 如果跟前面一個不符合，計算index到i之間的長度
6. result.push(長度)
7. 將result由大到小排列
8. 印出result[0]

實際上寫出來相當複雜
想請問這樣的解題邏輯
是否還有可以再更精簡、更直覺的方式
謝謝大家

```javascript
function solve(lines){

  let N = Number(lines[0])

  let arr = lines[1].split(' ')

  let result = []

  let index = 0

  for(let i=1; i<N; i++){
      
    if(i === N-1 && arr[i] !== arr[i-1]){

      result.push(i-index)

      result.push(1)

    }else if(i === N-1 && arr[i] === arr[i-1]){

      result.push(i-index+1)

    }else if (arr[i] !== arr[i-1]){

      result.push(i-index)

      index = i

    }else {

      continue 

  }

}

result.sort(function bigger (a, b) {
  return b - a
})

if(N === 1){
  
  console.log(1)
  
}else {
  
  console.log(result)

}

}

```

**Comment by allenliaooooooo** [comment]
應該可以用一個 `count` 存目前這個平台的長度，`result` 存最長平台的長度，然後一個 for loop 過去
數字一樣的時候就把 `count` +1，數字不一樣的時候比較 `result` 跟 `count`，取比較大的那個存進 `result`，把 `count` reset 回 0，繼續，直到結束

**Comment by HedgehogKUCC** [comment]
我的想法是先如何取出 `1 1 2 2 2 3 3 3 4 4` 這裡面 **最大的值** 來當作第一層 `for loop` 的條件

第二層 `for loop` 用 `1 1 2 2 2 3 3 3 4 4` 轉成 Number[] 後的長度當條件

根據第一層 `i = 1` 開始去跑第二層迴圈

如果相同就 `sum++` 跑完第二層迴圈時，把 `answer.push(sum)`

最後在 `Math.max(...answer)` 就可以找出長度最長的平台的長度

參考：[Math.max() - JavaScript | MDN](https://developer.mozilla.org/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/Math/max)

［程式碼］

```javascript
function solve(lines) {
  const numbers = lines[1].split(' ').map(item => Number(item));
  const maxNumber = Math.max(...numbers);
  const answer = [];
  
  for (let i = 1; i <= maxNumber; i++) {
    let sum = 0;
    for (let j = 0; j < numbers.length; j++) {
      if (i === numbers[j]) {
        sum++;
      }
    }
    answer.push(sum);
  }

  console.log(Math.max(...answer));
}
```
